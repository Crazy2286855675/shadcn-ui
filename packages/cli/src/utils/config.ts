import os from "os"
import path from "path"
import fs from "fs-extra"
import { z } from "zod"

// $HOME/.config/shadcn-ui/config.json
const CONFIG_PATH = path.join(
  os.homedir(),
  ".config",
  "shadcn-ui",
  "config.json"
)

const configSchema = z.object({
  projects: z.record(z.string(), z.string()),
})

type Config = z.infer<typeof configSchema>

// initialize the config
// create the config file if it doesn't exist
/**
 * ---
 * Initialize the config. Ensure that the config file exists and has the corret content
 */
export const init = async () => {
  await fs.ensureFile(CONFIG_PATH)

  const content = await fs.readFile(CONFIG_PATH, "utf-8")

  if (!!content) return
  await fs.writeJson(CONFIG_PATH, { projects: {} }, { spaces: 2 })
}

/**
 * ---
 * Get the config object
 */
export const get = async (): Promise<Config> => {
  const json = await fs.readJson(CONFIG_PATH)
  return configSchema.parse(json)
}

/**
 * ---
 * Upsert a project preference
 *
 * @param project - The current working dir
 * @param value - The path where to store the components
 */
export const upsertProject = async (project: string, value: string) => {
  const config = await get()

  config.projects[project] = value

  await fs.writeJSON(CONFIG_PATH, config, { spaces: 2 })
}
