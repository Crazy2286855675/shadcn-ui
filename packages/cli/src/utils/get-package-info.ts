import path from "path"
import fs from "fs-extra"
import { type PackageJson } from "type-fest"

export type ShadcnUIConfig = {
  "shadow-ui"?: {
    componentsDir?: string
  }
}

export type PkgJson = PackageJson & ShadcnUIConfig

export function getPackageInfo() {
  const packageJsonPath = path.join("package.json")

  return fs.readJSONSync(packageJsonPath) as PkgJson
}
