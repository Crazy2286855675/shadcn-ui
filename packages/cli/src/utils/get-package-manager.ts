import { promises as fs } from "fs"
import path from "path"

async function fileExists(path: string) {
  try {
    await fs.access(path)
    return true
  } catch {
    return false
  }
}

export async function getPackageManager(
  cwd = "."
): Promise<"yarn" | "pnpm" | "npm"> {
  // match based on lockfile
  const [yarnLock, npmLock, pnpmLock] = await Promise.all([
    fileExists(path.resolve(cwd, "yarn.lock")),
    fileExists(path.resolve(cwd, "package-lock.json")),
    fileExists(path.resolve(cwd, "pnpm-lock.yaml")),
  ])

  if (yarnLock) {
    return "yarn"
  }

  if (pnpmLock) {
    return "pnpm"
  }

  if (npmLock) {
    return "npm"
  }

  // match based on used package manager
  const userAgent = process.env.npm_config_user_agent

  if (!userAgent) {
    return "npm"
  }

  if (userAgent.startsWith("yarn")) {
    return "yarn"
  }

  if (userAgent.startsWith("pnpm")) {
    return "pnpm"
  }

  return "npm"
}
