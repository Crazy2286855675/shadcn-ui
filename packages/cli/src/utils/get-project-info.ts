import { existsSync } from "fs"
import path from "path"
import fs from "fs-extra"

export interface ProjectInfo {
  tsconfig: any | null
  alias: null | string
  srcDir: boolean
  appDir: boolean
}

export async function getProjectInfo() {
  const info: ProjectInfo = {
    tsconfig: null,
    alias: null,
    srcDir: false,
    appDir: false,
  }

  try {
    const tsconfig = await getTsConfig()
    const paths = tsconfig?.compilerOptions?.paths
    const alias = paths ? Object.keys(paths)[0].replace("*", "") : null

    return {
      tsconfig,
      alias,
      srcDir: existsSync(path.resolve("./src")),
      appDir:
        existsSync(path.resolve("./app")) ||
        existsSync(path.resolve("./src/app")),
    }
  } catch (error) {
    return info
  }
}

export async function getTsConfig() {
  try {
    const tsconfigPath = path.join("tsconfig.json")
    const tsconfig = await fs.readJSON(tsconfigPath)

    if (!tsconfig) {
      throw new Error("tsconfig.json is missing")
    }

    return tsconfig
  } catch (error) {
    return null
  }
}
