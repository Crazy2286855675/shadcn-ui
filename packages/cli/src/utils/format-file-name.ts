import { CASE_CONVENTION } from "@/src/utils/registry"
import toCamelCase from "lodash.camelcase"
import toKebabCase from "lodash.kebabcase"
import toSnakeCase from "lodash.snakecase"


export function formatFileName(
  fullFileName: string,
  caseConvention: CASE_CONVENTION
) {
  let fileNameWithExt = fullFileName;
  let fileName = fileNameWithExt;
  let path = '';
  let ext = '';

  if (fullFileName.includes('/')) {
    const lastSlashIndex = fullFileName.lastIndexOf('/');
    fileNameWithExt = fullFileName.substring(lastSlashIndex + 1);
    path = fullFileName.substring(0, lastSlashIndex + 1);
  }

  if (fileNameWithExt.includes('.')) {
    const lastDotIndex = fileNameWithExt.lastIndexOf('.');
    fileName = fileNameWithExt.substring(0, lastDotIndex);
    ext = fileNameWithExt.substring(lastDotIndex);
  }

  switch (caseConvention) {
    case CASE_CONVENTION.KEBAB:
      fileName = toKebabCase(fileName)
      break
    case CASE_CONVENTION.PASCAL:
      const camelCase = toCamelCase(fileName)
      fileName = camelCase.charAt(0).toUpperCase() + camelCase.slice(1)
      break
    case CASE_CONVENTION.CAMEL:
      fileName = toCamelCase(fileName)
      break
    case CASE_CONVENTION.SNAKE:
      fileName = toSnakeCase(fileName)
      break
  }

  return `${path}${fileName}${ext}`
}
