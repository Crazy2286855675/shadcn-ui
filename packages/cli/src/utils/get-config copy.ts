import { promises as fs } from "fs"
import path from "path"
import { logger } from "@/src/utils/logger"
import chalk from "chalk"
import { cosmiconfig } from "cosmiconfig"
import ora from "ora"
import prompts from "prompts"
import * as z from "zod"

export const DEFAULT_COMPONENTS = "components"
export const DEFAULT_UI = "components/ui"
export const DEFAULT_UTILS = "lib/utils.ts"
export const DEFAULT_STYLES = "app/globals.css"
export const DEFAULT_IMPORT_ALIAS = "@"
export const DEFAULT_TAILWIND_CONFIG = "tailwind.config.js"

// TODO: Figure out if we want to support all cosmiconfig formats.
// A simple components.json file would be nice.
const explorer = cosmiconfig("components", {
  searchPlaces: ["components.json"],
})

const configSchema = z
  .object({
    components: z.string().default(DEFAULT_COMPONENTS),
    ui: z.string().default(DEFAULT_UI),
    utils: z.string().default(DEFAULT_UTILS),
    styles: z.string().default(DEFAULT_STYLES),
    tailwindConfig: z.string().default(DEFAULT_TAILWIND_CONFIG),
    importAlias: z.string().default(DEFAULT_IMPORT_ALIAS),
  })
  .strict()

export type Config = z.infer<typeof configSchema>

export async function getConfig(search: string) {
  const config = await getRawConfig(search)

  if (!config) {
    return null
  }

  return await resolveConfigPaths(search, config)
}

export async function resolveConfigPaths(targetDir: string, config: Config) {
  return {
    ...config,
    components: path.resolve(targetDir, config.components),
    ui: path.resolve(targetDir, config.ui),
    utils: path.resolve(targetDir, config.utils),
    tailwindConfig: path.resolve(targetDir, config.tailwindConfig),
    styles: path.resolve(targetDir, config.styles),
  }
}

export async function getRawConfig(targetDir: string) {
  try {
    const configResult = await explorer.search(targetDir)

    if (!configResult) {
      return null
    }

    return configSchema.parse(configResult.config)
  } catch (error) {
    throw new Error(
      `Invald configuration found in ${targetDir}/components.json.`
    )
  }
}

export async function promptForConfig(targetDir: string) {
  const highlight = (text: string) => chalk.cyan(text)

  const options = await prompts([
    {
      type: "text",
      name: "components",
      message: `Where would you like to install ${highlight("components")}?`,
      initial: DEFAULT_COMPONENTS,
    },
    {
      type: "text",
      name: "ui",
      message: `Where would you like to install ${highlight("UI components")}?`,
      initial: DEFAULT_UI,
    },
    {
      type: "text",
      name: "utils",
      message: `Where should we place ${highlight(
        "utility helpers"
      )} such as cn()?`,
      initial: DEFAULT_UTILS,
    },
    {
      type: "text",
      name: "styles",
      message: `Where should we place ${highlight("Tailwind CSS styles")}?`,
      initial: DEFAULT_STYLES,
    },
    {
      type: "text",
      name: "importAlias",
      message: `Which ${highlight("import alias")} would you like to use?`,
      initial: DEFAULT_IMPORT_ALIAS,
    },
    {
      type: "text",
      name: "tailwindConfig",
      message: `Where is your ${highlight("tailwind.config.js")} located?`,
      initial: DEFAULT_TAILWIND_CONFIG,
    },
  ])

  const config = configSchema.parse(options)

  const { proceed } = await prompts({
    type: "confirm",
    name: "proceed",
    message: `Write configuration to ${highlight("components.json")}. Proceed?`,
    initial: true,
  })

  if (!proceed) {
    process.exit(0)
  }

  // Write to file.
  logger.info("")
  const spinner = ora(`Writing components.json...`).start()
  const targetPath = path.resolve(targetDir, "components.json")
  await fs.writeFile(targetPath, JSON.stringify(config, null, 2), "utf8")
  spinner.succeed()

  return resolveConfigPaths(targetDir, config)
}
