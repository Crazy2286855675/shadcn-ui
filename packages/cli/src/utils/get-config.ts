import { promises as fs } from "fs"
import path from "path"
import { logger } from "@/src/utils/logger"
import { resolveImport } from "@/src/utils/resolve-import"
import chalk from "chalk"
import { cosmiconfig } from "cosmiconfig"
import ora from "ora"
import prompts from "prompts"
import { loadConfig } from "tsconfig-paths"
import * as z from "zod"

export const DEFAULT_UI = "@/components/ui"
export const DEFAULT_UTILS = "@/lib/utils.ts"
export const DEFAULT_STYLES = "app/globals.css"
export const DEFAULT_TAILWIND_CONFIG = "tailwind.config.js"

// TODO: Figure out if we want to support all cosmiconfig formats.
// A simple components.json file would be nice.
const explorer = cosmiconfig("components", {
  searchPlaces: ["components.json"],
})

export const rawConfigSchema = z
  .object({
    importPaths: z.object({
      styles: z.string().default(DEFAULT_STYLES),
      "utils:cn": z.string().default(DEFAULT_UTILS),
      "components:ui": z.string().default(DEFAULT_UI),
    }),
    tailwindConfig: z.string().default(DEFAULT_TAILWIND_CONFIG),
  })
  .strict()

export type RawConfig = z.infer<typeof rawConfigSchema>

export const configSchema = rawConfigSchema.extend({
  resolvedPaths: z.object({
    styles: z.string(),
    "utils:cn": z.string(),
    "components:ui": z.string(),
  }),
})

export type Config = z.infer<typeof configSchema>

export async function getConfig(cwd: string) {
  const config = await getRawConfig(cwd)

  if (!config) {
    return null
  }

  return await resolveConfigPaths(cwd, config)
}

export async function resolveConfigPaths(cwd: string, config: RawConfig) {
  // Read tsconfig.json.
  const tsConfig = await loadConfig(cwd)

  if (tsConfig.resultType === "failed") {
    throw new Error(
      `Failed to load tsconfig.json. ${tsConfig.message ?? ""}`.trim()
    )
  }

  return configSchema.parse({
    ...config,
    tailwindConfig: path.resolve(cwd, config.tailwindConfig),
    resolvedPaths: {
      styles: await resolveImport(config.importPaths.styles, tsConfig),
      "utils:cn": await resolveImport(config.importPaths["utils:cn"], tsConfig),
      "components:ui": await resolveImport(
        config.importPaths["components:ui"],
        tsConfig
      ),
    },
  })
}

export async function getRawConfig(cwd: string): Promise<RawConfig | null> {
  try {
    const configResult = await explorer.search(cwd)

    if (!configResult) {
      return null
    }

    return rawConfigSchema.parse(configResult.config)
  } catch (error) {
    throw new Error(`Invald configuration found in ${cwd}/components.json.`)
  }
}
