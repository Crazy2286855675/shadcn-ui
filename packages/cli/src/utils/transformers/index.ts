import { promises as fs } from "fs"
import { tmpdir } from "os"
import path from "path"
import { Config } from "@/src/utils/get-config"
import { transformRsc } from "@/src/utils/transformers/transform-rsc"
import { Project, type SourceFile } from "ts-morph"

import { transformImport } from "./transform-import"

export type Transformer = (opts: {
  sourceFile: SourceFile
  filename: string
  raw: string
  config: Config
}) => Promise<SourceFile>

const transformers: Transformer[] = [transformImport, transformRsc]

const project = new Project()

async function createTempSourceFile(filename: string) {
  const dir = await fs.mkdtemp(path.join(tmpdir(), "shadcn-"))
  return path.join(dir, filename)
}

export async function transform(filename: string, raw: string, config: Config) {
  const tempFile = await createTempSourceFile(filename)
  const sourceFile = project.createSourceFile(tempFile, raw)

  for (const transformer of transformers) {
    transformer({ sourceFile, filename, raw, config })
  }

  return sourceFile.getFullText()
}
