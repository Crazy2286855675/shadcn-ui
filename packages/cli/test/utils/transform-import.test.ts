import { expect, test } from "vitest"

import { transform } from "../../src/utils/transformers"

test("transform import", async () => {
  expect(
    await transform(
      "test.ts",
      `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/new-york/ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/new-york/box"

    import { cn } from "@/lib/utils"
    `,
      {
        aliases: {
          components: "@/components",
          utils: "@/lib/utils",
        },
      }
    )
  ).toMatchSnapshot()

  expect(
    await transform(
      "test.ts",
      `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/new-york/ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/new-york/box"

    import { cn, foo, bar } from "@/lib/utils"
    import { bar } from "@/lib/utils/bar"
    `,
      {
        aliases: {
          components: "~/src/components",
          utils: "~/lib",
        },
      }
    )
  ).toMatchSnapshot()

  expect(
    await transform(
      "test.ts",
      `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/new-york/ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/new-york/box"

    import { cn } from "@/lib/utils"
    import { bar } from "@/lib/utils/bar"
    `,
      {
        aliases: {
          components: "~/src/components",
          utils: "~/src/utils",
        },
      }
    )
  ).toMatchSnapshot()
})

test("transform rsc", async () => {
  expect(
    await transform(
      "test.ts",
      `import * as React from "react"
import { Foo } from "bar"
    `,
      {
        rsc: true,
      }
    )
  ).toMatchSnapshot()

  expect(
    await transform(
      "test.ts",
      `"use client"

      import * as React from "react"
import { Foo } from "bar"
    `,
      {
        rsc: true,
      }
    )
  ).toMatchSnapshot()

  expect(
    await transform(
      "test.ts",
      `"use client"

      import * as React from "react"
import { Foo } from "bar"
    `,
      {
        rsc: false,
      }
    )
  ).toMatchSnapshot()

  expect(
    await transform(
      "test.ts",
      `"use foo"

      import * as React from "react"
import { Foo } from "bar"

"use client"
    `,
      {
        rsc: false,
      }
    )
  ).toMatchSnapshot()
})
