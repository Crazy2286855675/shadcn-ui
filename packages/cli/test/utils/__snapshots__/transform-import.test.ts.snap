// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform import 1`] = `
"import * as React from \\"react\\"
import { Foo } from \\"bar\\"
    import { Button } from \\"@/components/ui/button\\"
    import { Label} from \\"ui/label\\"
    import { Box } from \\"@/components/box\\"

    import { cn } from \\"@/lib/utils\\"
    "
`;

exports[`transform import 2`] = `
"import * as React from \\"react\\"
import { Foo } from \\"bar\\"
    import { Button } from \\"~/src/components/ui/button\\"
    import { Label} from \\"ui/label\\"
    import { Box } from \\"~/src/components/box\\"

    import { cn, foo, bar } from \\"~/lib\\"
    import { bar } from \\"@/lib/utils/bar\\"
    "
`;

exports[`transform import 3`] = `
"import * as React from \\"react\\"
import { Foo } from \\"bar\\"
    import { Button } from \\"~/src/components/ui/button\\"
    import { Label} from \\"ui/label\\"
    import { Box } from \\"~/src/components/box\\"

    import { cn } from \\"~/src/utils\\"
    import { bar } from \\"@/lib/utils/bar\\"
    "
`;

exports[`transform import 4`] = `
"import * as React from \\"react\\"
import { Foo } from \\"bar\\"
    import { Button } from \\"~/src/components/button\\"
    import { Label} from \\"ui/label\\"
    import { Box } from \\"@/registry/new-york/box\\"

    import { cn } from \\"~/src/utils\\"
    import { bar } from \\"@/lib/utils/bar\\"
    "
`;

exports[`transform import 5`] = `
"import * as React from \\"react\\"
import { Foo } from \\"bar\\"
    import { Button } from \\"~/src/ui/button\\"
    import { Label} from \\"ui/label\\"
    import { Box } from \\"@/registry/new-york/box\\"

    import { cn } from \\"~/src/utils\\"
    import { bar } from \\"@/lib/utils/bar\\"
    "
`;

exports[`transform import 6`] = `
"import * as React from \\"react\\"
    import { AlertDialog } from \\"@/components/ui/AlertDialog\\"
    import { Toaster } from \\"@/components/ui/Toaster\\"
    import { useToast } from \\"@/components/ui/useToast\\"
    "
`;

exports[`transform import 7`] = `
"import * as React from \\"react\\"
      import {
        DropdownMenu,
        DropdownMenuContent,
        DropdownMenuItem,
        DropdownMenuLabel,
        DropdownMenuSeparator,
        DropdownMenuTrigger,
      } from \\"@/components/ui/dropdown_menu\\"
    "
`;
